File: code-java/src/main/java/me/yonghong/algo/lc/lc0101/Lc0102.java
package me.yonghong.algo.lc.lc0101;

import me.yonghong.algo.Solution;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

/**
 * 102. 二叉树的层序遍历
 *
 * @author yonghongwang#163.com
 * @link <a href="https://leetcode-cn.com/problems/binary-tree-level-order-traversal/"></a>
 * @link <a href="https://leetcode.com/problems/binary-tree-level-order-traversal/"></a>
 * @since 2021/12/1
 **/
public class Lc0102 implements Solution {

    public static void main(String[] args) {
        new Lc0102().test();
    }

    @Override
    public void test() {
        System.out.println(levelOrder(stringToTreeNode("[3,9,20,null,null,15,7]")));
        System.out.println(levelOrderDFS1(stringToTreeNode("[3,9,20,null,null,15,7]")));
        System.out.println(levelOrderDFS2(stringToTreeNode("[3,9,20,null,null,15,7]")));
        System.out.println(levelOrder(stringToTreeNode("[]")));
        System.out.println(levelOrderDFS1(stringToTreeNode("[]")));
        System.out.println(levelOrderDFS2(stringToTreeNode("[]")));
    }

    public List<List<Integer>> levelOrder(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }
        Deque<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        List<List<Integer>> res = new ArrayList<>();

        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> currentLevel = new ArrayList<>();
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.pop();
                currentLevel.add(node.val);
                if (node.left != null) {
                    queue.add(node.left);
                }
                if (node.right != null) {
                    queue.add(node.right);
                }
            }
            res.add(currentLevel);
        }

        return res;
    }

    public List<List<Integer>> levelOrderDFS1(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(root, res, 0);
        return res;
    }

    private void dfs(TreeNode root, List<List<Integer>> res, int level) {
        if (root == null) {
            return;
        }
        if (res.size() <= level) {
            res.add(new ArrayList<>());
        }
        res.get(level).add(root.val);
        dfs(root.left, res, level + 1);
        dfs(root.right, res, level + 1);
    }

    public List<List<Integer>> levelOrderDFS2(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        Deque<Pair<TreeNode, Integer>> stack = new LinkedList<>();
        stack.push(new Pair<>(root, 0));
        while (!stack.isEmpty()) {
            Pair<TreeNode, Integer> nodePair = stack.pop();
            Integer level = nodePair.second;
            if (res.size() <= level) {
                res.add(new ArrayList<>());
            }
            TreeNode node = nodePair.first;
            res.get(level).add(node.val);
            if (node.right != null) {
                stack.push(new Pair<>(node.right, level + 1));
            }
            if (node.left != null) {
                stack.push(new Pair<>(node.left, level + 1));
            }
        }
        return res;
    }

    class Pair<F, S> {
        F first;
        S second;

        public Pair(F first, S second) {
            this.first = first;
            this.second = second;
        }
    }
}

File: code-python/algo/lc/lc0101/lc0102.py
import os
import sys
import collections
from typing import List

sys.path.append(os.path.dirname(os.path.dirname(sys.path[0])))
from base.solution import Solution, TreeNode

# 102. 二叉树的层序遍历
# https://leetcode-cn.com/problems/binary-tree-level-order-traversal/
# https://leetcode.com/problems/binary-tree-level-order-traversal/


class Lc0101(Solution):

  def levelOrder(self, root: TreeNode) -> List[List[int]]:
    # BFS 广度优先
    if not root:
      return []

    result = []
    queue = collections.deque()
    queue.append(root)

    while queue:
      level_size = len(queue)
      current_level = []
      for _ in range(level_size):
        node = queue.popleft()
        current_level.append(node.val)
        if node.left:
          queue.append(node.left)
        if node.right:
          queue.append(node.right)
      result.append(current_level)
    return result

  def levelOrderDFS1(self, root):
    # DFS 深度优先
    result = []
    self.dfs(root, result, 0)
    return result

  def dfs(self, root, result, level):
    if not root:
      return
    if len(result) <= level:
      result.append([])
    result[level].append(root.val)
    self.dfs(root.left, result, level + 1)
    self.dfs(root.right, result, level + 1)

  def levelOrderDFS2(self, root):
    # DFS 深度优先
    result = []
    if not root:
      return result

    stack = collections.deque()
    stack.append([root, 0])

    while stack:
      node_pair = stack.pop()
      node = node_pair[0]
      level = node_pair[1]
      if len(result) <= level:
        result.append([])
      result[level].append(node.val)
      if node.right:
        stack.append([node.right, level + 1])
      if node.left:
        stack.append([node.left, level + 1])
    return result

  def test(self):
    ret = self.levelOrder(self.stringToTreeNode('[3,9,20,null,null,15,7]'))
    print(self.int2dArrayToString(ret))
    ret = self.levelOrderDFS1(self.stringToTreeNode('[3,9,20,null,null,15,7]'))
    print(self.int2dArrayToString(ret))
    ret = self.levelOrderDFS2(self.stringToTreeNode('[3,9,20,null,null,15,7]'))
    print(self.int2dArrayToString(ret))
    ret = self.levelOrder(self.stringToTreeNode('[]'))
    print(self.int2dArrayToString(ret))
    ret = self.levelOrderDFS1(self.stringToTreeNode('[]'))
    print(self.int2dArrayToString(ret))
    ret = self.levelOrderDFS2(self.stringToTreeNode('[]'))
    print(self.int2dArrayToString(ret))


def main():
  Lc0101().test()


if __name__ == '__main__':
  main()
